Feature/cancel reservations job

MonikaLukowska opened the pull request:
This PR adds cancel outdated reservation cron job. Implementation includes:

- adding redis, sidekiq, sidekiq-cron
- adding job
-  cron job test
<img width="1395" alt="Zrzut ekranu 2022-10-5 o 19 29 43" src="https://user-images.githubusercontent.com/56922072/194124389-b86010ee-2538-462a-b97d-42030b251bb1.png">

MonikaLukowska committed:
feat:add sidekiq, sidekiq cron and redis

MonikaLukowska committed:
feat:add schedule

MonikaLukowska committed:
feat:add job schedule

MonikaLukowska committed:
add cancel outdated reservations job

MonikaLukowska committed:
sidekiq to procfile

MonikaLukowska committed:
add job spec

kodowicz commented on app/jobs/cancel_outdated_reservations_job.rb:
```diff
@@ -0,0 +1,27 @@
+  end
+
+  def reservations
+    Reservation.where(status: 'created')
```
```suggestion
    Reservation.where(status: Reservation::CREATED)
```

kodowicz commented on app/jobs/cancel_outdated_reservations_job.rb:
```diff
@@ -0,0 +1,27 @@
+
+  def perform
+    reservations.each do |reservation|
+      cancel_reservation(reservation, status).call unless outdated?(reservation)
```
you don't need to pass `status` here :D it can be called in `cancel_reservation` method :)

```ruby
  def cancel_reservation(reservation)
    Reservations::UseCases::Update.new(reservation: reservation, status: status)
  end
  ```

MonikaLukowska replied:
it even is called in cancel_reservation🙈

kodowicz commented on app/jobs/cancel_outdated_reservations_job.rb:
```diff
@@ -0,0 +1,27 @@
+  end
+
+  def outdated?(reservation)
+    reservation.seance.start_time >= DateTime.now + 30.minutes
```
it would be nice to have this `30.minutes` as a constant in a `Reservation` model :)

kodowicz commented on config/schedule.yml:
```diff
@@ -0,0 +1,3 @@
+cancel_reservation_job:
+  cron: '*/5 * * * *'
+  class: 'CancelOutdatedReservationsJob'
```
Why do you want to create a cron job instead of calling a job and scheduling it to 30 minutes? In this case, it could be a situation when the cron job was performed 1 minute ago and now I'm buying a ticket so it will be checked when the ticket was created 29 minutes ago and next time when it was created 34 minutes ago. The ticket should have been canceled for 4 minutes. 

Definitely a better solution is to schedule a job for 30minutes when a reservation has been created.

rrothenberger replied:
Because I told them that better solution is to handle that as cron job 😄

Reasoning for that:
- Reservations should be canceled 30 minutes before the screening starts, not 30 minutes after booking
- This has pretty nasty consequences - if we have few screenings at the same time, it means there could be maybe oven over thousand reservations that each will have job queued for the same time, which could block sidekiq for some duration, as that 1000+ jobs need to execute. Because those will execute for all reservations, not only not paid ones. Cron job will run only one and handle all jobs
- it handles edge cases better - eg. if time of screening will change. Let's say that somebody bought ticket a week earlier for some big premiere. The ticket was for first screening in the morning, at 7:00 AM. But then two days later we will learn that Tauron will be doing something and we will not have electricity till 8 AM. So we are moving this screening to 9AM. Now cron job will handle this change without any issue, but job scheduled at time of reservation will not - it will still run at 6:30 AM.
- Cron jobs better handle errors - if for some reason there will be issue with DB for 10 minutes, scheduled job will exhaust all retries and fail never canceling the tickets. Cron job on the other hand will still cancel them, just 10 minutes later.
- yes, there are edge cases with screenings set for times like 7:32 - but that's unlikely to be actual screening time, as most cinemas try to set them for at least times divisible by 5. And it can be fixed by just running cron job each minutes if it would be needed.

kodowicz replied:
thanks for the explanation!

MonikaLukowska replied:
Now after reading your comments one more issue crossed my mind. If not confirmed reservations should be cancelled 30 min before seance, then client shouldn't be allowed to create reservation <=30 min. before seance, right?

marcel-strzalka replied:
I think, there is no other way. If you allow your users to create reservations 30 minutes before show start it will lead to some nasty consequences: Someone will make a reservation and next minute it will be cancelled or they might even buy them if they are quick enough and app doesn't protect against that.

rrothenberger replied:
Yeah, users can not create reservations when seanse is less than 30 minutes away, but employes should be able to :smile:

kodowicz commented on spec/jobs/cancel_outdated_reservations_job_spec.rb:
```diff
@@ -0,0 +1,22 @@
+  end
+
+  describe '#perform_now' do
+    let(:reservation) { create(:reservation) }
```
I know you have a default value for a `status` column but I can't see it looking at the factory for this variable so for a better readability or changes in a future you could define status like this.
```suggestion
    let(:reservation) { create(:reservation, status: Reservation::CREATED) }
```

kodowicz replied:
You could even create `trait` that would tell more what reservation you are creating (because when status is created you would like to have other columns with a specific data)
```ruby
FactoryBot.define do
  factory :reservation do
  trait :created do
    status { Reservation::CREATED }
  end
end
```

kodowicz commented on spec/jobs/cancel_outdated_reservations_job_spec.rb:
```diff
@@ -0,0 +1,22 @@
+    before { reservation }
+
+    it 'cancels outdated reservation' do
+      expect { described_class.perform_now }.to change { reservation.reload.status }.from('created').to('cancelled')
```
Another thing - you created constants so use them :D 
```suggestion
      expect { described_class.perform_now }
        .to change { reservation.reload.status }
        .from(Reservation::CREATED)
        .to(Reservation::CANCELED)
```

MonikaLukowska committed:
ref:apply CR remarks

create confirmation_deadline constant, use status constants, remove useless argument

MonikaLukowska committed:
test:refactor according to cr remarks

add created trait to reservation factory and use it in job spec

marcel-strzalka commented on app/jobs/cancel_outdated_reservations_job.rb:
```diff
@@ -0,0 +1,27 @@
+  end
+
+  def outdated?(reservation)
+    reservation.seance.start_time >= DateTime.now + Reservation::CONFIRMATION_DEADLINE
```
I think it should look a little bit different:

You basically need to check if it is already 30 minutes before the seance.

```ruby
DateTime.now >= reservation.seance.start_time - Reservation::CONFIRMATION_DEADLINE
```

MonikaLukowska replied:
I think that in your version it should be rather 
`DateTime.now <= reservation.seance.start_time - Reservation::CONFIRMATION_DEADLINE`

marcel-strzalka replied:
Things like that can be quite confusing and difficult to reason about, so to make it clearer I decided to draw a picture, hope it helps. 

![Screenshot 2022-10-06 at 12 18 53](https://user-images.githubusercontent.com/113047354/194289021-40a2be51-6430-472a-a83e-bdaacc7cdc52.png)

MonikaLukowska replied:
your're right, thanks!

MonikaLukowska committed:
fix:fix logic in job

MonikaLukowska committed:
chore:move sidekiq config to initializers

MonikaLukowska committed:
feat:add time restriction to create reservation use case

MonikaLukowska committed:
test:update tests

MonikaLukowska committed:
fix: fix logic in use case and spec

MonikaLukowska committed:
fix:setup app time zone

MonikaLukowska committed:
fix: user confirmable - update spec, factories seeds

MonikaLukowska committed:
style:remove repeated redis gem

MonikaLukowska committed:
fix:prevent from creating invalid reservation

MonikaLukowska committed:
test:check if reservation not created when invalid

MonikaLukowska committed:
fix:correct params

MonikaLukowska committed:
perf:avoind N+1

MonikaLukowska committed:
test:refactor expectations
